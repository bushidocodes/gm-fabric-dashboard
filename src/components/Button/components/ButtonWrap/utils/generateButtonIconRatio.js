/**
 * Helper function that maps button sizes to numbers representing icon ratios
 *
 * @param {string} size
 * @returns number
 */
function mapSizeToButtonIconRatio(size) {
  switch (size) {
    case "xs":
      return 0.5;
    case "sm":
      return 0.75;
    case "lg":
      return 1.25;
    case "xl":
      return 1.5;
    case "normal":
    default:
      return 1;
  }
}

/**
 * Utility function that generates a CSS mixin containing button icon rulesets
 *
 * @param {string} size
 * @returns string
 */
function generateButtonIconRatio(size) {
  const viewboxSize = 24; // default viewbox size 24px
  const ratio = mapSizeToButtonIconRatio(size); // default ratio 1
  const viewboxDimensions = viewboxSize * ratio;
  return `> svg { width: ${/* sc-value */ viewboxDimensions}px; 
  height: ${/* sc-value */ viewboxDimensions}px; }`;
}

export default generateButtonIconRatio;
