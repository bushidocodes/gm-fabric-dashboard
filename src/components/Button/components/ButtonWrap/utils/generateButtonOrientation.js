import { spacingScale } from "style/styleFunctions";

/**
 * Utility function that generates a CSS mixin containing button orientation rulesets
 *
 * @param {string} [alignment="horizontal"]
 * @param {number} [spacingBase=0.5]
 * @returns string
 */
function generateButtonOrientation(
  alignment = "horizontal",
  spacingBase = 0.5
) {
  if (alignment === "horizontal") {
    return `
      flex-direction: row;
      > :not(:first-child) {
        margin-left: ${/* sc-value */ spacingScale(spacingBase)};
        flex-basis: auto;
      }
    `;
  } else if (alignment === "vertical") {
    return `
      flex-direction: column;
      padding-top: ${/* sc-value */ spacingScale(spacingBase)};
      > :not(:first-child) {
        margin-top: ${/* sc-value */ spacingBase / 2}px;
      }
    `;
  }
}

export default generateButtonOrientation;
