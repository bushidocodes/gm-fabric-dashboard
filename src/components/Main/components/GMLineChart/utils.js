// importing a single function instead of the whole _
import isEmpty from "lodash/isEmpty";

/**
 * Converts timeseries with time stamps to a string of values without timestamps
 *
 * @param {array} array a timeseries array with time stamps
 * @returns {array}     an array of values
 */
export function numericalTimeSeriesFunc(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    return array[0].map(cell => cell[1]);
  } else {
    return 0;
  }
}

/**
 * Produces an average of array of values
 *
 * @param {array} array an array of values
 * @returns {number}
 */
export function average(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    return array.reduce((a, b) => a + b) / array.length;
  } else {
    return 0;
  }
}

/**
 * Produces a median of array of values
 * (if no middle value, average of 2 most central values is returned)
 *
 * @param {array} array an array of values
 * @returns {number}
 */
export function median(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    const len = array.length;
    let median = 0;
    array.sort();

    if (len % 2 === 0) {
      // if even, average of two middle numbers
      median = (array[len / 2 - 1] + array[len / 2]) / 2;
    } else {
      // if odd, middle number only
      median = array[(len - 1) / 2];
    }
    return median;
  } else {
    return 0;
  }
}

/**
 * Produces a mode(s) of array of values
 * if more than one mode, returns all applicable modes as well as their frequency
 *
 * @param {array} array an array of values
 * @returns {array}     an array that contains mode(s) and frequency value
 */
export function modes(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    let modes = [];
    let count = [];
    let i;
    let number;
    let maxIndex = 0;

    for (let i = 0; i < array.length; i++) {
      number = array[i];
      count[number] = (count[number] || 0) + 1;
      if (count[number] > maxIndex) {
        maxIndex = count[number];
      }
    }

    for (i in count)
      if (count.hasOwnProperty(i)) {
        if (count[i] === maxIndex) {
          modes.push(Number(i));
        }
      }

    modes.push({ frequency: maxIndex });

    if (maxIndex > 1) {
      return modes;
    } else {
      modes = "no value occurs more than once";
      return modes;
    }
  } else {
    return 0;
  }
}

/**
 * Produces a string that contains descriptive statistics of the graph
 *
 * @param {array} timeSeries an array of values
 * @param {string} title chart title
 * @returns {string}     returns a string with summary chart statistics
 */
export function screenReaderGraphDescription(timeSeries, title, intl) {
  const numericalTimeSeries = numericalTimeSeriesFunc(timeSeries);

  // We don't have negative values in our charts, thus average of 0 means that all values are 0, aka no meaningful data
  if (average(numericalTimeSeries) > 0) {
    let screenReaderGraphData = {
      median: median(numericalTimeSeries).toFixed(2),
      average: average(numericalTimeSeries).toFixed(2),
      mode: modes(numericalTimeSeries),
      max: numericalTimeSeries.sort()[numericalTimeSeries.length - 1],
      min: numericalTimeSeries.sort()[0],
      dataPoints: numericalTimeSeries.length,
      timeSeries: numericalTimeSeries,
      title
    };

    if (
      typeof screenReaderGraphData.mode !== "string" &&
      screenReaderGraphData.mode.length > 1
    ) {
      screenReaderGraphData = Object.assign({}, screenReaderGraphData, {
        mode: `Mode has frequency of ${
          screenReaderGraphData.mode.pop().frequency
        } and values of ${screenReaderGraphData.mode}`
      });
    }

    return intl.formatMessage(
      {
        id: "GMLineChart.screenReaderGraphDescription.withData",
        defaultMessage:
          "A tabular representation of the {title} chart data: median {median} average {average} mode {mode} maximum {max} minimum {min} number of observations {dataPoints} complete data time series follows {timeSeries}",
        description: "Screen reader description for GMLineChart"
      },
      screenReaderGraphData
    );
  } else {
    return intl.formatMessage({
      id: "GMLineChart.screenReaderGraphDescription.noData",
      defaultMessage: `The average for currently displayed data is equal to 0.`,
      description: "Screen reader description for GMLineChart"
    });
  }
}
