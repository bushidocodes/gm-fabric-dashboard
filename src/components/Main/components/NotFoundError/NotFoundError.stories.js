import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, text } from "@storybook/addon-knobs/react";

import NotFoundError from "./NotFoundError";

storiesOf("NotFoundError", module)
  .addDecorator(withKnobs)
  .add("default", () => (
    <div style={{ display: "flex" }}>
      <NotFoundError errorMsg={text("errorMsg")} />
    </div>
  ));
