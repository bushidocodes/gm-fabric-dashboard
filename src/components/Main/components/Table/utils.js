import React from "react";
import TableColHeader from "components/Main/components/TableColHeader";
import TableColLatencyHeader from "components/Main/components/TableColLatencyHeader";

/** function getTableHeaders
 * takes Table type and returns appropriate headers with styling
 * for Function/Routes Table or GMServiceTable (Instances Table)
 * @param {object} items
 * @returns {React.element}
 */
export const getTableHeaders = type => {
  let headerNames = [];
  if (type === "Instance") headerNames = [type, "Uptime"];
  if (type === "Route" || type === "Function")
    headerNames = [type, "Requests/sec", "Requests", "Error %", "Latency"];
  let styleProps = {};

  // apply greater flex-basis for first column (20%)
  // requests/sec is a sparkline header and set to align left
  // set text-align to right for additional text columns
  return headerNames.map((header, headerIdx) => {
    styleProps = {};
    if (headerIdx === 0) styleProps = { flex: "1 1 20%" };
    if (headerIdx !== 0 && header !== "Requests/sec")
      styleProps = { textAlign: "right" };

    if (header === "Latency") {
      return <TableColLatencyHeader key={header} />;
    } else {
      return (
        <TableColHeader style={styleProps} key={header}>
          {header}
        </TableColHeader>
      );
    }
  });
};

// simple helper that evaluates and returns route or func props
export const getItem = (route, func) => {
  if (route !== undefined) return route;
  if (func !== undefined) return func;
};
