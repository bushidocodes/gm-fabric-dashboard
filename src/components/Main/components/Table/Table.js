import { PropTypes } from "prop-types";
import React from "react";
import _ from "lodash";

import TableLineItem from "./components/TableLineItem";
import GMServiceTableLineItem from "./components/GMServiceTableLineItem";
import UpTime from "components/UpTime";
import ArrayValue from "components/ArrayValue";
import TableDisplay from "components/Main/components/TableDisplay";
import TableHeader from "components/Main/components/TableHeader";
import TableBody from "components/Main/components/TableBody";
import { relativeReqPercent } from "utils";
// Table-specific utils functions
import { getTableHeaders, getItem } from "./utils";

Table.propTypes = {
  items: PropTypes.array,
  serviceIsMetered: PropTypes.bool,
  serviceName: PropTypes.string,
  serviceVersion: PropTypes.string,
  status: PropTypes.string,
  type: PropTypes.string
};

/** Renders JVM > Routes table, Go > Functions table, Go > Routes table, Instance Service table
 * Includes a Header and maps 0..n rows of functions data to TableLineItems
 * @export
 * @param {any}
 * @returns
 */
export default function Table({
  items,
  type,
  serviceName,
  serviceVersion,
  serviceIsMetered,
  status
}) {
  // adds relativeReqPercent field to items for viz-fill-bar rendering to Route/Function table
  if (type === "Route" || type === "Function")
    items = relativeReqPercent(items, "requests");

  return (
    <TableDisplay>
      <TableHeader>{getTableHeaders(type)}</TableHeader>
      <TableBody>
        {(type === "Route" || type === "Function") &&
          items.map(
            ({
              func,
              route,
              verb,
              errorPercent,
              inThroughput,
              outThroughput,
              latency50,
              latency99,
              relativeReqPercent,
              requests,
              requestsPerSecond_dygraph,
              requestsPerSecond_sparkline
            }) => (
              <TableLineItem
                errorPercent={errorPercent}
                item={getItem(func, route)}
                key={_.replace(`${func}${route}/${verb}`, "undefined", "")}
                latency50={latency50}
                latency99={latency99}
                relativeReqPercent={relativeReqPercent}
                requests={requests}
                requestsPerSecond_dygraph={requestsPerSecond_dygraph}
                requestsPerSecond_sparkline={requestsPerSecond_sparkline}
                verb={verb}
              />
            )
          )}
        {type === "Instance" &&
          items.map(instance => (
            <GMServiceTableLineItem
              name={instance.name}
              serviceIsMetered={serviceIsMetered}
              uptime={
                <UpTime
                  startTime={instance.start_time}
                  render={uptime => (
                    <ArrayValue>
                      {_.map(uptime, el => <span key={el}>{el} </span>)}
                    </ArrayValue>
                  )}
                />
              }
              path={`/${serviceName}/${serviceVersion}/${instance.name}`}
              status={status}
              key={`${serviceName}/${serviceVersion}/${instance.name}`}
            />
          ))}
      </TableBody>
    </TableDisplay>
  );
}
