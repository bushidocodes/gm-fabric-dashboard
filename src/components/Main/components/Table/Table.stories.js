import React from "react";
import { storiesOf } from "@storybook/react";
import StoryRouter from "storybook-router";
import { withKnobs, object } from "@storybook/addon-knobs/react";

import Table from "./index.js";

const mockRoutesTableItems = [
  {
    errorPercent: "0.000",
    errorsCount: 0,
    inThroughput: 0,
    latency50: 0,
    latency99: 9,
    outThroughput: 1580,
    requests: 17725,
    requestsPerSecond_dygraph: {
      data: [
        [new Date("2017-07-18T22:13:34.314Z"), 0],
        [new Date("2017-07-18T22:13:49.215Z"), 0]
      ],
      attributes: ["time", "Requests"]
    },
    requestsPerSecond_sparkline: [0, 0],
    route: "/functionalroles",
    verb: "GET"
  },
  {
    errorPercent: "0.000",
    errorsCount: 0,
    inThroughput: 0,
    latency50: 0,
    latency99: 0,
    outThroughput: 0,
    requests: 1,
    requestsPerSecond_dygraph: {
      data: [
        [new Date("2017-07-18T22:13:34.314Z"), 0],
        [new Date("2017-07-18T22:13:49.215Z"), 0]
      ],
      attributes: ["time", "Requests"]
    },
    requestsPerSecond_sparkline: [0, 0],
    route: "/ping",
    verb: "GET"
  }
];

const mockFuncsTableItems = [
  {
    errorPercent: "0.000",
    errorsCount: 0,
    func: "CatalogStream",
    inThroughput: 227,
    latency50: 1956,
    latency90: 3469,
    latency95: 3484,
    latency99: 3484,
    latency9990: 3484,
    latency9999: 3484,
    latencyAvg: 1942.692308,
    latencyCount: 13,
    latencyMax: 3484,
    latencyMin: 600,
    outThroughput: 2889,
    requests: 18,
    requestsPerSecond_dygraph: {
      data: [
        [new Date("2017-07-18T22:13:34.314Z"), 0],
        [new Date("2017-07-18T22:13:49.215Z"), 0]
      ],
      attributes: ["time", "Requests"]
    },
    requestsPerSecond_sparkline: [0, 0]
  },
  {
    errorPercent: "0.000",
    errorsCount: 0,
    func: "OrderItem",
    inThroughput: 225,
    latency50: 2339,
    latency90: 3476,
    latency95: 3565,
    latency99: 3565,
    latency9990: 3565,
    latency9999: 3565,
    latencyAvg: 2398.615385,
    latencyCount: 13,
    latencyMax: 3565,
    latencyMin: 1345,
    outThroughput: 143,
    requests: 13,
    requestsPerSecond_dygraph: {
      data: [
        [new Date("2017-07-18T22:13:34.314Z"), 0],
        [new Date("2017-07-18T22:13:49.215Z"), 0]
      ],
      attributes: ["time", "Requests"]
    },
    requestsPerSecond_sparkline: [0, 0]
  }
];

const mockInstanceTableItems = [
  { name: "serviceID1", start_time: 600000000000 },
  { name: "serviceID2", start_time: 300000000000 },
  { name: "serviceID3", start_time: 400000000000 }
];

storiesOf("Table", module)
  .addDecorator(withKnobs)
  .addDecorator(StoryRouter())
  .add("Routes Table", () => {
    return (
      <div style={{ width: "100%" }}>
        <Table type="Route" items={object("items", mockRoutesTableItems)} />
      </div>
    );
  })
  .add("Functions Table", () => {
    return (
      <div style={{ width: "100%" }}>
        <Table type="Function" items={object("items", mockFuncsTableItems)} />
      </div>
    );
  })
  .add("Instance Table", () => {
    return (
      <div style={{ width: "100%" }}>
        <Table
          status="stable"
          type="Instance"
          items={object("items", mockInstanceTableItems)}
        />
      </div>
    );
  });
