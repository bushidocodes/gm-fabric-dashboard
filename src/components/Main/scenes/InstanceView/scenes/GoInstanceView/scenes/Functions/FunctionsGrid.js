import { PropTypes } from "prop-types";
import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import _ from "lodash";

import Table from "components/Main/components/Table";
import TableToolbar from "components/Main/components/TableToolbar";
import { getFunctionsTable } from "utils/go/selectors";
import ErrorBoundary from "components/ErrorBoundary";
import NotFoundError from "components/Main/components/NotFoundError";
import withUrlState from "components/withUrlState";

/**
 * Go Functions Container
 * Parent of FunctionsTable and FunctionsTableToolbar
 * Contains sort and filter logic for FunctionsTable
 * @class FunctionsGrid
 * @extends {Component}
 */

class FunctionsGrid extends Component {
  static propTypes = {
    funcs: PropTypes.array,
    setUrlState: PropTypes.func.isRequired,
    urlState: PropTypes.object.isRequired
  };
  /**
   *  Helper function that takes the funcs passed as props and sorts according to how sortByAttribute and ascending is set
   * in the url bar.
   * @param {Array} funcs
   */
  sort = routes => {
    const {
      ascending = "true",
      sortByAttribute = "func"
    } = this.props.urlState;
    const sortOrder = ascending === "true" ? ["asc"] : ["desc"];

    return _.orderBy(routes, sortByAttribute, sortOrder);
  };

  /**
   * Helper function used internally to either sort by the key if not yet used to sort or toggle ascending / descending
   * if the key is already active.
   *  When sorting by latency and error percent and the sortKey is new, set the initial sort order to descending
   * @param {String} newSortByAttribute
   */
  setSortByAttribute = newSortByAttribute => {
    const {
      urlState: { ascending = "true", sortByAttribute = "func" },
      setUrlState
    } = this.props;

    if (sortByAttribute === newSortByAttribute) {
      setUrlState({
        ascending: !JSON.parse(ascending)
      });
    } else if (
      newSortByAttribute === "errorPercent" ||
      newSortByAttribute.includes("latency")
    ) {
      setUrlState({
        sortByAttribute: newSortByAttribute,
        ascending: false
      });
    } else {
      setUrlState({
        sortByAttribute: newSortByAttribute
      });
    }
  };

  render() {
    const {
      setUrlState,
      urlState: { filterString = "", sortByAttribute = "func" },
      funcs
    } = this.props;

    if (funcs && funcs.length > 0) {
      return (
        <Fragment>
          <TableToolbar
            searchInputProps={{
              filterString,
              setFilterString: filterString => setUrlState({ filterString }),
              searchPlaceholder: "Search Functions"
            }}
            sortByProps={{
              sortByOptions: [
                {
                  value: "func",
                  label: "Function"
                },
                {
                  value: "requests",
                  label: "Requests"
                },
                {
                  value: "errorCount",
                  label: "Error %"
                },
                {
                  value: "latency50",
                  label: "Latency 50%"
                },
                {
                  value: "latency99",
                  label: "Latency 99%"
                }
              ],
              sortByAttribute,
              setSortByAttribute: this.setSortByAttribute
            }}
          />
          <ErrorBoundary>
            <Table
              type={"Function"}
              items={this.sort(
                this.props.funcs.filter(
                  funcObj =>
                    funcObj.func
                      .toLowerCase()
                      .indexOf(filterString.trim().toLowerCase()) !== -1
                )
              )}
            />
          </ErrorBoundary>
        </Fragment>
      );
    } else {
      return <NotFoundError errorMsg={"No Functions Found"} />;
    }
  }
}

function mapStateToProps(state) {
  return {
    funcs: getFunctionsTable(state)
  };
}

export default connect(mapStateToProps)(withUrlState()(FunctionsGrid));
