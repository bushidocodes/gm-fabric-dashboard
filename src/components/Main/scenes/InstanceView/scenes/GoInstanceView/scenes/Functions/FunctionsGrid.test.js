import React from "react";
import configureStore from "redux-mock-store";
import _ from "lodash";

import * as noFuncState from "json/mockReduxState";
import * as state from "json/mockReduxStateMetrics";
import FunctionsGrid from "./index";
import NotFoundError from "components/Main/components/NotFoundError";
import { mountWithIntl, renderWithIntl } from "utils/i18nTesting";

//import Action effects
import "services";
import Table from "components/Main/components/Table";

const mockStore = configureStore();
const mockState = state.default,
  noMetricsState = noFuncState.default;

// Filter out metrics starting with the key function
noMetricsState.instance.metrics = _.omitBy(
  noFuncState.default.instance.metrics,
  (value, key) => key.substr(0, 8) === "function"
);
let wrapper;

const FunctionsGridWithMockStore = (
  <FunctionsGrid store={mockStore(mockState)} />
);
const FunctionsGridWithMissingMetricsStore = (
  <FunctionsGrid store={mockStore(noMetricsState)} />
);

const sortByOptions = [
  {
    value: "func",
    label: "Function"
  },
  {
    value: "requests",
    label: "Requests"
  },
  {
    value: "errorCount",
    label: "Error %"
  },
  {
    value: "latency50",
    label: "Latency 50%"
  },
  {
    value: "latency99",
    label: "Latency 99%"
  }
];

describe("Go Instance Functions View: <FunctionsGrid/>", () => {
  test("Matches snapshot", () => {
    const tree = renderWithIntl(FunctionsGridWithMockStore);
    expect(tree).toMatchSnapshot();
  });

  // FunctionsGridWithMissingMetricsStore does not contain any functions data
  test("returns NotFoundError if no functions are found ", () => {
    wrapper = mountWithIntl(FunctionsGridWithMissingMetricsStore);
    expect(wrapper.find(NotFoundError).length).toBe(1);
  });

  test("returns correct number of <Table> and does not render <NotFoundError> when functions are found ", () => {
    wrapper = mountWithIntl(FunctionsGridWithMockStore);
    expect(wrapper.find(Table).length).toBe(1);
    expect(wrapper.find(NotFoundError).length).toBe(0);
  });
});

describe("FunctionsGrid Child Components", () => {
  beforeEach(() => {
    wrapper = mountWithIntl(FunctionsGridWithMockStore);
  });

  test("passes props to TableToolbar", () => {
    expect(wrapper.find("TableToolbar").props()).toMatchObject({
      searchInputProps: {
        filterString: "",
        searchPlaceholder: "Search Functions"
      },
      sortByProps: {
        sortByOptions,
        sortByAttribute: "func"
      }
    });
  });

  test("passes functions as props to Table", () => {
    expect(wrapper.find(Table).props()).toMatchObject({
      items: [
        {
          errorPercent: "0",
          errorsCount: 0,
          func: "CatalogStream",
          inThroughput: 227,
          latency50: 1956,
          latency90: 3469,
          latency95: 3484,
          latency99: 3484,
          latency9990: 3484,
          latency9999: 3484,
          latencyAvg: 1942.692308,
          latencyCount: 13,
          latencyMax: 3484,
          latencyMin: 600,
          outThroughput: 2889,
          requests: 13
        },
        {
          errorPercent: "0",
          errorsCount: 0,
          func: "OrderItem",
          inThroughput: 225,
          latency50: 2339,
          latency90: 3476,
          latency95: 3565,
          latency99: 3565,
          latency9990: 3565,
          latency9999: 3565,
          latencyAvg: 2398.615385,
          latencyCount: 13,
          latencyMax: 3565,
          latencyMin: 1345,
          outThroughput: 143,
          requests: 13
        }
      ]
    });
  });
});
