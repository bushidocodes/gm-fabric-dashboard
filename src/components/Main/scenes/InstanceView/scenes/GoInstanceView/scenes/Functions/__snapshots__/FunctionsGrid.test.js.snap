// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Go Instance Functions View: <FunctionsGrid/> Matches snapshot 1`] = `
<WithUrlStateWrapper
  dispatch={[Function]}
  funcs={
    Array [
      Object {
        "errorPercent": "0.000",
        "errorsCount": 0,
        "func": "CatalogStream",
        "inThroughput": 227,
        "latency50": 1956,
        "latency90": 3469,
        "latency95": 3484,
        "latency99": 3484,
        "latency9990": 3484,
        "latency9999": 3484,
        "latencyAvg": 1942.692308,
        "latencyCount": 13,
        "latencyMax": 3484,
        "latencyMin": 600,
        "outThroughput": 2889,
        "requests": 13,
        "requestsPerSecond_dygraph": Array [
          Array [
            Array [
              2017-11-14T22:48:14.557Z,
              0,
            ],
            Array [
              2017-11-14T22:48:19.598Z,
              0,
            ],
            Array [
              2017-11-14T22:48:24.508Z,
              0,
            ],
            Array [
              2017-11-14T22:48:29.508Z,
              0,
            ],
            Array [
              2017-11-14T22:48:34.507Z,
              0,
            ],
            Array [
              2017-11-14T22:48:39.507Z,
              0,
            ],
          ],
          Object {
            "labels": Array [
              "time",
              "Requests",
            ],
          },
        ],
        "requestsPerSecond_sparkline": Array [
          0,
          0,
          0,
          0,
          0,
        ],
      },
      Object {
        "errorPercent": "0.000",
        "errorsCount": 0,
        "func": "OrderItem",
        "inThroughput": 225,
        "latency50": 2339,
        "latency90": 3476,
        "latency95": 3565,
        "latency99": 3565,
        "latency9990": 3565,
        "latency9999": 3565,
        "latencyAvg": 2398.615385,
        "latencyCount": 13,
        "latencyMax": 3565,
        "latencyMin": 1345,
        "outThroughput": 143,
        "requests": 13,
        "requestsPerSecond_dygraph": Array [
          Array [
            Array [
              2017-11-14T22:48:14.557Z,
              0,
            ],
            Array [
              2017-11-14T22:48:19.598Z,
              0,
            ],
            Array [
              2017-11-14T22:48:24.508Z,
              0,
            ],
            Array [
              2017-11-14T22:48:29.508Z,
              0,
            ],
            Array [
              2017-11-14T22:48:34.507Z,
              0,
            ],
          ],
          Object {
            "labels": Array [
              "time",
              "Requests",
            ],
          },
        ],
        "requestsPerSecond_sparkline": Array [
          0,
          0,
          0,
          0,
        ],
      },
    ]
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
